account/src/lib.rs

```diff
@@ -1,80 +1,56 @@
-use near_sdk::{env, near, store::LookupSet, AccountId, BorshStorageKey, PanicOnDefault};
+use defuse_contracts::account::AccountContract;
+use near_contract_standards::non_fungible_token::NonFungibleToken;
+use near_sdk::{env, near, AccountId, BorshStorageKey, PanicOnDefault};
 
-use self::account_db::AccountDb;
-
-mod account_db;
+mod nft;
 
 #[derive(BorshStorageKey)]
 #[near(serializers=[borsh])]
 enum Prefix {
-    Accounts,
-    Indexers,
+    OwnerById,
+    Metatada,
+    Enumeration,
+    Approvals,
 }
 
 #[near(contract_state)]
 #[derive(PanicOnDefault)]
 pub struct AccountContractImpl {
-    owner_id: AccountId,
-    /// MPC contract id.
-    mpc_contract_id: AccountId,
-    /// List of indexers. Accounts which allow to add a new account.
-    indexers: LookupSet<AccountId>,
-    /// Key here is `account_id + '.' + derivation_path`
-    /// Value is account id of the owner.
-    accounts: AccountDb,
-}
-
-#[near]
-impl AccountContract for AccountContractImpl {
-    fn create_account(&mut self, account_id: AccountId, derivation_path: String) {
-        // Only indexers can call this transaction.
-        let predecessor_id = env::predecessor_account_id();
-        self.assert_indexer(&predecessor_id);
-
-        self.accounts
-            .add_account(account_id, derivation_path, Account::default())
-            .unwrap();
-    }
-
-    fn change_owner(&mut self, from: &AccountId, to: AccountId, derivation_path: String) {
-        self.accounts
-            .change_owner(from, to, derivation_path)
-            .unwrap();
-    }
-
-    fn get_accounts(&self, account_id: &AccountId) -> Vec<(String, Account)> {
-        self.accounts.get_accounts(account_id).unwrap()
-    }
-
-    fn mpc_contract(&self) -> &AccountId {
-        &self.mpc_contract_id
-    }
+    accounts: NonFungibleToken,
+    // TODO: MPC
 }
 
 #[near]
 impl AccountContractImpl {
     #[init]
     #[must_use]
     #[allow(clippy::use_self)]
-    pub fn new(owner_id: AccountId, mpc_contract_id: AccountId) -> Self {
+    pub fn new(owner: AccountId) -> Self {
         Self {
-            owner_id,
-            mpc_contract_id,
-            indexers: LookupSet::new(Prefix::Indexers),
-            accounts: AccountDb::new(Prefix::Accounts),
+            accounts: NonFungibleToken::new::<_, u8, u8, u8>(
+                Prefix::OwnerById,
+                owner,
+                None,
+                None,
+                None,
+            ),
         }
     }
+}
 
-    #[private]
-    pub fn set_mpc_contract(&mut self, contract_id: AccountId) {
-        self.mpc_contract_id = contract_id;
-    }
-
-    fn assert_indexer(&self, account_id: &AccountId) {
-        assert!(
-            self.indexers.contains(account_id),
-            "Only indexers allow adding an account"
+#[near]
+impl AccountContract for AccountContractImpl {
+    #[payable]
+    // TODO: make owner optional and default to env::predecessor_account_id()?
```

Comment:
I guess an intent contract could create an account. In such case `env::predecessor_account_id()` is not a case.

Reply:
why?

Reply:
Because we discussed it.

Reply:
Because we discussed it.

Reply:
As in https://github.com/defuse-is-near/defuse-contracts/pull/8#discussion_r1658458877, we would need to discuss format of `derivation_path`. As an example:
> Derivation path `user_1.intent_shard_1 ` can mean following:
> This derivation path was generated by `user_1` and is designed to be "claimed" only by `intent_shard_1`.
> Account shard can implement the logic which allows to "claim" this derivation path if and only if the initiator of the intent is `user_1`. But the logical "owner" of this derivation path would be `intent_shard_1`, so we can make sure that this derivation path was never used to sign any transaction.